### ISSUE TITLE
[New azd-template] todo-java-postgresql-terraform

### ISSUE THREAD

**Issue by fangjian0423** [Azure/awesome-azd#76]
## New azd-template Submission

Fill in the sections below to populate a new Gallery carry for your template. An example of a valid template object is given at the end for reference. **All fields are mandatory unless explicitly identified as optional**

### 1. Title 
_Short template name that will be used in card title on gallery_

**ToDo Application - Java PostgreSQL AppService Terraform**

## 2. Description
_1-2 sentence description of the architecture or solution that is defined by the template_

**A complete ToDo application that includes everything you need to build, deploy, and monitor an Azure solution. This application uses the Azure Developer CLI (azd) to get you up and running on Azure quickly, React.js for the Web application, Java for the API, Azure Database for PostgreSQL - Flexible Server for storage, and Azure Monitor for monitoring and logging.**

## 3. Author Profile
_A link to the contributor's GitHub profile (ideal) or other profile link (for identification_.

**https://github.com/fangjian0423**

## 4. Author Name
_The name to be shown in the grey button on the card - will be linked profile website above_.

**Jimmy Fang**

## 5. Template Repo
_The GitHub repo containing the template being contributed - must be a valid, complete template_

**https://github.com/fangjian0423/todo-java-postgresql-terraform**

## 6. Link to PR where you made your repo `azd` compatible
_To allow us to provide feedback on your template, please include a link to the PR where you added `azd` compatibility_

It's a new project. Could you pls create a new issue if you want to provide feedback.

## 7. Standard Tags
_One or more tags representing the template. Look at the currently available tags on the Gallery page and pick at least 1 tag for programming language used, and 1 tag for Azure services integrated_

**`java`, `appservice`, `postgresql`, `monitor`, `passwordless`, `application insights`, `terraform`** 

## 7. Additional Tags (`optional`)
_Don't see a relevant standard tag for your template? Request these tags be added, and then used for your template. These could be new languages, services, or application domains_.

**`spring`**

## 8. Architecture Diagram or Application Screenshot
_Used as default image for template card - must be representative of solution being enabled. Submit image by adding it as file upload to this issue_

![image](https://github.com/fangjian0423/todo-java-postgresql-terraform/blob/main/assets/resources.png?raw=true)

---

## Example Of Valid Template

```
{
  title: 'ToDo Application - Java PostgreSQL AppService',
  description:
    'Complete React.js ToDo application, Java for the API, Azure Database for PostgreSQL - Flexible Server for storage, and Azure Monitor for monitoring and logging.',
    preview: `todo-java-postgresql.png`,
  website: 'https://github.com/fangjian0423',
  author: 'Jimmy Fang',
  source: 'https://github.com/fangjian0423/todo-java-postgresql-terraform',
  tags: ['java', 'appservice', 'postgresql', 'monitor', 'passwordless', 'application insights', 'terraform']
}
```
---

## Next Steps

Once submitted the issue will be reviewed - we plan to do reviews on a rolling basis at regular intervals. The process will involve checking that template information is complete, and the template is valid. 
 * If we have questions or enhancements, we will add comments in issue thread (**issue stays open**)
 * If the contribution is approved, we'll update the gallery directly (**issue will then be closed**)

**Comment by weikanglim** [comment]
@fangjian0423 Sorry for the delay, I'm starting to take a look at this template now.

**Comment by weikanglim** [comment]
@fangjian0423 Getting an error that doesn't resolve after retrying. Do you know what might be happening?

```
 Error: creating Flexible Server (Subscription: "faa080af-****-40ad-9cce-e1a450ca5b57"
│ Resource Group Name: "rg-weilim-java-postgres-tf"
│ Server Name: "psqlf-yjjinjyynzkzm"): polling after Create: Code="InternalServerError" Message="An unexpected error occured while processing the request. Tracking ID: 'dfbf691a-5a54-4174-9a0d-aee7b7bdfd09'"
│
│   with module.postgresql.azurerm_postgresql_flexible_server.psqlServer,
│   on modules\postgresql\postgresql.tf line 35, in resource "azurerm_postgresql_flexible_server" "psqlServer":
│   35: resource "azurerm_postgresql_flexible_server" "psqlServer" {
│
╵

ERROR: deployment failed: error deploying infrastructure: template Deploy failed:  , err:failed running terraform apply:  (exit status 1)
```

**Comment by fangjian0423** [comment]
Hi @weikanglim , it seems this is an issue in postgresql, I find https://github.com/Azure/azure-dev/issues/933 talking about it.

I run successfully after retrying 2 times (location is eastus2), and didn't get the error again in the subsequent trying. Could you pls try it again?

BTW, I meet another error:

```
module.loganalytics.azurerm_log_analytics_workspace.workspace: Creating...
╷
│ Error: performing CreateOrUpdate: workspaces.WorkspacesClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=<nil> Code="InvalidPolicyParameters" Message="A function or parameter in policy assignment 'acat-custom-policy-set-assignment' associated with the policy definition 'log-analytics-workspace-retention' could not be validated. Please either fix the policy or remove the policy assignment to unblock. If using template functions, try following the tips in: https://aka.ms/policy-avoiding-template-failures. The inner error is 'The template language function 'length' expects its parameter to be an array, object, or a string. The provided value is of type 'Integer'. Please see https://aka.ms/arm-template-expressions/#length for usage details.'."
│
│   with module.loganalytics.azurerm_log_analytics_workspace.workspace,
│   on modules/loganalytics/loganalytics.tf line 23, in resource "azurerm_log_analytics_workspace" "workspace":
│   23: resource "azurerm_log_analytics_workspace" "workspace" {
│
```

Do you get this error?

I get the same error when run `azd up` with `todo-python-mongo-terraform` template.

So I created an issue for this error: https://github.com/hashicorp/terraform-provider-azurerm/issues/20197

**Comment by weikanglim** [comment]
@fangjian0423 Thanks for the pointer. Choosing a different location than `eastus2` did make it progress further.

I ended up hitting a different issue. This time it relates to `local-exec`, where it fails to execute on Windows, because the default interpreter isn't `sh`/`bash`. `terraform` doesn't provide native ability to conditionally specify scripts, but I did find this [resource](https://github.com/hashicorp/terraform/issues/23785) which hints at how we could solve this in a different manner. I would recommend creating a similar powershell script for Windows to bootstrap the AD roles. Alternatively, you could require in Readme a shell interpreter to be present, but the user should not need to modify the terraform script to make it work. (Could be an env variable if you preferred). Note that in the up-and-coming `azd` feature for hooks, `postprovision` scripts can be specified per OS. 

Related to that, it looks like the script would require `psql` to be present on the user's machine? Can you add this as a pre-req?

**Comment by fangjian0423** [comment]
@weikanglim I have updated the template.

1. Modify `local-exec` to support Windows. Thanks for your provided terraform resource 😄
2. Add PostgreSQL as a prerequisite in README

Could you please pull the newest code to try again?

> Note that in the up-and-coming azd feature for hooks, postprovision scripts can be specified per OS.

Awesome, looking forward to see it.

**Comment by rajeshkamal5050** [comment]
@v-xuto can you do a dry on this template and update here for any issues/feedback?

**Comment by v-xuto** [comment]
@fangjian0423 We have tested the template `todo-java-postgresql-terraform`, we encountered the following issues：
- When we run the command `azd up --template https://github.com/fangjian0423/todo-java-postgresql-terraform`, and use the default location `eastus2`, we encountered a postgresql issue like https://github.com/Azure/azure-dev/issues/933.  But there is a difference, no matter how many times we run it, it will fail.
- Executing the `azd up` command is successful when we use the location `eastus`, but there is an issue when opening the web and api url. We have filed an issue https://github.com/fangjian0423/todo-java-postgresql-terraform/issues/1, please review.
- When we run the command `azd pipeline config`, there is a regression issue `azd pipeline started failing on Github actions`, it has been filed under azure-dev, issue link: https://github.com/Azure/azure-dev/issues/1532.

**Comment by v-xuto** [comment]
@fangjian0423 
About the issue [#933](https://github.com/Azure/azure-dev/issues/933) `Can't deploy environment that includes a PostGreSQL flexible server in eastus2`, we can't test this template in location `eastus2`, because our subscription `Azure Data Developer Playground` is not allowed to provision in `Eastus2` for `PostGreSQL flexible server`.

Now, when executing `azd pipeline config` fails, we have filed a new issue https://github.com/fangjian0423/todo-java-postgresql-terraform/issues/2. Please review.

**Comment by v-xuto** [comment]
@jongio We have finished testing this template, and there is only one issue(https://github.com/fangjian0423/todo-java-postgresql-terraform/issues/3) now. 
After discussing with the author, this issue should be a service issue (Azure Database for PostgreSQL - Flexible Server is not stable for the AAD admin setting in terraform), it's not a template issue.
Regarding this [issue3](https://github.com/fangjian0423/todo-java-postgresql-terraform/issues/3), how should we deal with it?

**Comment by rajeshkamal5050** [comment]
> After discussing with the author, this issue should be a service issue (Azure Database for PostgreSQL - Flexible Server is not stable for the AAD admin setting in terraform), it's not a template issue.

@jongio @fangjian0423 call out in the template readme as a known issue on the service?

@jongio @weikanglim  please review.

**Comment by fangjian0423** [comment]
The postgresql flexible server AAD admin setting in terraform is not stable. The error code is "InternalServerError", very confuse.

Retry can resolve it.

I will describe the case on readme **Troubleshooting** section.

**Comment by jongio** [comment]
@weikanglim - Where are we at with adding this to azure-dev repo?

**Comment by fangjian0423** [comment]
hi @rajeshkamal5050 @jongio @weikanglim , any update about this template?

**Comment by wbreza** [comment]
I also ran into the AAD issue upon deployment.  I would recommend we turn off AAD integration on the template if possible.

**Comment by fangjian0423** [comment]
> I would recommend we turn off AAD integration on the template if possible.

Thanks for your suggestion. I won't turn off it because this template want to promote [Azure Active Directory Authentication with PostgreSQL Flexible Server](https://learn.microsoft.com/en-us/azure/postgresql/flexible-server/concepts-azure-ad-authentication).

**Comment by rajeshkamal5050** [comment]
@jongio @savannahostrowski is it okay for this to be served from https://github.com/fangjian0423/todo-java-postgresql-terraform?

**Comment by rajeshkamal5050** [comment]
Closing this in favor of - https://github.com/Azure/azure-dev/pull/1792
